### PR TITLE
Remove wrappers around storage-related listers in the predicate

### LINKED ISSUE #84185: Remove wrappers around storage-related listers in the scheduler
**What would you like to be added**:
Remove wrappers around storage-related listers in the scheduler: https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithm/predicates/predicates.go#L161-L228

predicates should just use shared informer listers directly.

**Why is this needed**:
They provide no value, at the same time removing them simplifies the scheduler framework.

/sig scheduling
/priority important-soon
/assign @hprateek43

### PR DESCRIPTION
**Pull request by zouyee**
**What type of PR is this?**

/kind cleanup
/kind feature

**What this PR does / why we need it**:
Remove wrappers around storage-related listers in the scheduler, it provide no value, at the same time removing them simplifies the scheduler framework.

**Which issue(s) this PR fixes**:

Fixes #84185

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/scheduler/algorithm/predicates/csi_volume_predicate.go [hunk 1]
```diff
@@ -23,6 +23,9 @@ import (
 	storagev1beta1 "k8s.io/api/storage/v1beta1"
 	"k8s.io/apimachinery/pkg/util/rand"
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	storagelisters "k8s.io/client-go/listers/storage/v1"
+	v1beta1storagelisters "k8s.io/client-go/listers/storage/v1beta1"
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithm/predicates/csi_volume_predicate.go]: Please do NOT import from client-go, import from `k8s.io/api/...`

File: pkg/scheduler/algorithm/predicates/csi_volume_predicate.go [hunk 2]
```diff
@@ -43,10 +46,10 @@ type InTreeToCSITranslator interface {
...
 // CSIMaxVolumeLimitChecker defines predicate needed for counting CSI volumes
 type CSIMaxVolumeLimitChecker struct {
-	csiNodeInfo CSINodeInfo
-	pvInfo      PersistentVolumeInfo
-	pvcInfo     PersistentVolumeClaimInfo
-	scInfo      StorageClassInfo
+	csiNodeInfo v1beta1storagelisters.CSINodeLister
+	pvInfo      corelisters.PersistentVolumeLister
+	pvcInfo     corelisters.PersistentVolumeClaimLister
+	scInfo      storagelisters.StorageClassLister
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithm/predicates/csi_volume_predicate.go]: Suggest to rename them to: csiNodeLister, pvLister, pvcLister, scLister.

**Comment by zouyee** [comment]
@Huang-Wei @ahg-g PTAL

**Comment by ahg-g** [comment]
You are going to have to rebase after https://github.com/kubernetes/kubernetes/pull/84193 is merged, so I will review after that happens.

**Comment by ahg-g** [comment]
#84193 is merged, can you please rebase.

**Comment by zouyee** [comment]
@ahg-g has rebased

File: pkg/scheduler/listers/fake/listers.go [hunk 3]
```diff
@@ -250,9 +254,49 @@ func (pvcs PersistentVolumeClaimInfo) GetPersistentVolumeClaimInfo(namespace str
...
+	for i := range pvcs {
+		ps[i] = &pvcs[i]
+	}
+	return &persistentVolumeClaimNamespaceLister{
+		pvcs:      ps,
+		namespace: namespace,
+	}
+}
+
+var _ corelisters.NodeLister = NodeLister{}
```
Comment (ahg-g) [inline pkg/scheduler/listers/fake/listers.go]: please remove this

File: pkg/scheduler/listers/fake/listers.go [hunk 4]
```diff
@@ -250,9 +254,49 @@ func (pvcs PersistentVolumeClaimInfo) GetPersistentVolumeClaimInfo(namespace str
...
+func (nodes NodeLister) List(selector labels.Selector) (ret []*v1.Node, err error) {
+	return nil, fmt.Errorf("not implemented")
+}
+
+// ListWithPredicate allows custom methods to be added to
+// NodeLister.
+func (nodes NodeLister) ListWithPredicate(predicate corelisters.NodeConditionPredicate) ([]*v1.Node, error) {
+	return nil, fmt.Errorf("not implemented")
+}
+
```
Comment (ahg-g) [inline pkg/scheduler/listers/fake/listers.go]: remove all of this as well, just do the storage-related listers.

File: pkg/scheduler/algorithm/predicates/csi_volume_predicate.go [hunk 5]
```diff
@@ -55,12 +58,12 @@ type CSIMaxVolumeLimitChecker struct {
 
 // NewCSIMaxVolumeLimitPredicate returns a predicate for counting CSI volumes
 func NewCSIMaxVolumeLimitPredicate(
-	csiNodeInfo CSINodeInfo, pvInfo PersistentVolumeInfo, pvcInfo PersistentVolumeClaimInfo, scInfo StorageClassInfo) FitPredicate {
+	csiNodeInfo v1beta1storagelisters.CSINodeLister, pvInfo corelisters.PersistentVolumeLister, pvcInfo corelisters.PersistentVolumeClaimLister, scInfo storagelisters.StorageClassLister) FitPredicate {
```
Comment (wgliang) [inline pkg/scheduler/algorithm/predicates/csi_volume_predicate.go]: Parameter names also need to be consistent *Lister.

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 6]
```diff
@@ -333,8 +264,8 @@ type VolumeFilter struct {
 // The predicate looks for both volumes used directly, as well as PVC volumes that are backed by relevant volume
 // types, counts the number of unique volumes, and rejects the new pod if it would place the total count over
 // the maximum.
-func NewMaxPDVolumeCountPredicate(filterName string, csiNodeInfo CSINodeInfo, scInfo StorageClassInfo,
-	pvInfo PersistentVolumeInfo, pvcInfo PersistentVolumeClaimInfo) FitPredicate {
+func NewMaxPDVolumeCountPredicate(filterName string, csiNodeInfo v1beta1storagelisters.CSINodeLister, scInfo storagelisters.StorageClassLister,
```
Comment (wgliang) [inline pkg/scheduler/algorithm/predicates/predicates.go]: ditoo: Parameter names also need to be consistent *Lister.

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 7]
```diff
@@ -700,11 +631,11 @@ type VolumeZoneChecker struct {
 // determining the zone of a volume during scheduling, and that is likely to
 // require calling out to the cloud provider.  It seems that we are moving away
 // from inline volume declarations anyway.
-func NewVolumeZonePredicate(pvInfo PersistentVolumeInfo, pvcInfo PersistentVolumeClaimInfo, classInfo StorageClassInfo) FitPredicate {
+func NewVolumeZonePredicate(pvInfo corelisters.PersistentVolumeLister, pvcInfo corelisters.PersistentVolumeClaimLister, classInfo storagelisters.StorageClassLister) FitPredicate {
```
Comment (wgliang) [inline pkg/scheduler/algorithm/predicates/predicates.go]: ditoo

File: pkg/scheduler/framework/plugins/nodevolumelimits/azure.go [hunk 8]
```diff
@@ -52,18 +52,10 @@ func (pl *AzureDiskLimits) Filter(ctx context.Context, _ *framework.CycleState,
...
-	pvInfo := &predicates.CachedPersistentVolumeInfo{
-		PersistentVolumeLister: informerFactory.Core().V1().PersistentVolumes().Lister(),
-	}
-	pvcInfo := &predicates.CachedPersistentVolumeClaimInfo{
-		PersistentVolumeClaimLister: informerFactory.Core().V1().PersistentVolumeClaims().Lister(),
-	}
-	classInfo := &predicates.CachedStorageClassInfo{
-		StorageClassLister: informerFactory.Storage().V1().StorageClasses().Lister(),
-	}
+	csiNodeInfo := informerFactory.Storage().V1beta1().CSINodes().Lister()
```
Comment (wgliang) [inline pkg/scheduler/framework/plugins/nodevolumelimits/azure.go]: ditoo

File: pkg/scheduler/framework/plugins/nodevolumelimits/cinder.go [hunk 9]
```diff
@@ -52,19 +52,10 @@ func (pl *CinderLimits) Filter(ctx context.Context, _ *framework.CycleState, pod
...
-	}
-	pvcInfo := &predicates.CachedPersistentVolumeClaimInfo{
-		PersistentVolumeClaimLister: informerFactory.Core().V1().PersistentVolumeClaims().Lister(),
-	}
-	classInfo := &predicates.CachedStorageClassInfo{
-		StorageClassLister: informerFactory.Storage().V1().StorageClasses().Lister(),
-	}
-
+	csiNodeInfo := informerFactory.Storage().V1beta1().CSINodes().Lister()
+	pvInfo := informerFactory.Core().V1().PersistentVolumes().Lister()
```
Comment (wgliang) [inline pkg/scheduler/framework/plugins/nodevolumelimits/cinder.go]: ditoo

**Review by wgliang — CHANGES_REQUESTED** [review]

**Comment by zouyee** [comment]
@ahg-g please take another look

File: pkg/scheduler/algorithm/predicates/csi_volume_predicate.go (hunk 5 above)
Comment (ahg-g) [reply pkg/scheduler/algorithm/predicates/csi_volume_predicate.go]: +1

**Review by ahg-g — COMMENTED** [review]
> @ahg-g please take another look

can you please fix the naming as @wgliang suggested.

File: pkg/scheduler/algorithm/predicates/csi_volume_predicate.go (hunk 5 above)
Comment (zouyee) [reply pkg/scheduler/algorithm/predicates/csi_volume_predicate.go]: Done.

**Comment by ahg-g** [comment]
/lgtm
/approve

In a followup PR, do you mind changing the fake listers names from PersistentVolumeClaimInfo/CSINodeInfo/PersistentVolumeInfo/StorageClassInfo to PersistentVolumeClaimLister/CSINodeLister/PersistentVolumeLister/StorageClassLister

**Comment by zouyee** [comment]
@ahg-g ok

**Pull request merged**